viewer/packages/data-providers/src/image-360-data-providers/descriptor-providers/datamodels/cdm/Cdf360CdmDescriptorProvider.ts

```diff
@@ -0,0 +1,251 @@
+
+    assert(images.length === fileDescriptors.length, 'Expected each 360 image to have 6 faces');
+
+    const imagesGroupedWithFileDescriptors = zip(images, fileDescriptors)
+      .filter(
+        (imageWithFileInfo): imageWithFileInfo is [ImageInstanceResult, FileInfo[]] =>
+          imageWithFileInfo[0] !== undefined && imageWithFileInfo[1] !== undefined
+      )
+      .map(([image, fileDescriptors]) => ({ image, fileDescriptors }));
+
+    const [imagesWithoutStation, imagesWithStation] = partition(imagesGroupedWithFileDescriptors, image => {
+      return image.image.properties.cdf_cdm['Cognite360Image/v1'].station360 === undefined;
+    });
+
+    const groups = groupBy(imagesWithStation, imageResult => {
+      const station = imageResult.image.properties.cdf_cdm['Cognite360Image/v1'].station360 as InstanceIdentifier;
+      return `${station.externalId}-${station.space}`;
+    });
+
+    const ret = Object.values(groups)
+      .concat(imagesWithoutStation.map(p => [p]))
+      .map(imageWithFileDescriptors => {
+        return this.getHistorical360ImageSet(collectionId, collectionLabel, imageWithFileDescriptors);
+      });
+
+    return ret;
+  }
+
+  private async queryCollection({
+    image360CollectionExternalId,
+    space
+  }: Image360DataModelIdentifier): Promise<ExhaustedQueryResult> {
+    const result: ExhaustedQueryResult = {
+      image_collection: [],
+      images: [],
+      stations: []
+    };
+
+    const query = get360CollectionQuery(image360CollectionExternalId, space);
+
+    const imageLimit = query.with.images.limit;
+
+    let nextCursor: QueryNextCursors<Cdf360FdmQuery> | undefined = undefined;
+    let hasNext = true;
+
+    while (hasNext) {
+      const {
+        image_collection,
+        images,
+        stations,
+        nextCursor: currentCursor
+      }: QueryResult = await this._dmsSdk.queryNodesAndEdges(query, nextCursor);
+      if (result.image_collection.length === 0) {
+        result.image_collection.push(...image_collection);
+      }
+      result.images.push(...images);
+      result.stations.push(...stations);
+
+      hasNext =
+        images.length === imageLimit && currentCursor?.images !== undefined && currentCursor?.stations !== undefined;
+      nextCursor = {
+        images: currentCursor?.images,
+        stations: currentCursor?.stations
+      };
+    }
+
+    return result;
+  }
+
+  private async getFileDescriptors(images: ImageResult) {
+    const imageProps = images.map(image => image.properties.cdf_cdm['Cognite360Image/v1']);
+    const cubeMapFileIds = imageProps.flatMap(imageProp => [
+      imageProp.front as DirectRelationReference,
+      imageProp.back as DirectRelationReference,
+      imageProp.left as DirectRelationReference,
+      imageProp.right as DirectRelationReference,
+      imageProp.top as DirectRelationReference,
+      imageProp.bottom as DirectRelationReference
+    ]);
+
+    const externalIdBatches = chunk(chunk(cubeMapFileIds, 1000), 15);
```

Comment:
What is 15 and can add it into Constant variable

Reply:
I'm not sure about why it's 15, but I would guess that it has something to do about the File API's concurrency limits. I asked Christopher who wrote the datamodeling provider I based this CDM implementation on, and he did not remember why 15 was chosen